test/03_device-state.ts

```diff
@@ -510,6 +553,314 @@ export default () => {
 									api_heartbeat_state: DeviceOnlineStates.Offline,
 								});
 							});
+
+							describe('when the write cache expires while the device is on Timeout (eg API stops serving requests)', function () {
+								let device3: fakeDevice.Device;
+								const device3ChangeEventSpy = sinon.spy();
+
+								before(async function () {
+									device3 = await fakeDevice.provisionDevice(
+										admin,
+										applicationId,
+									);
+									stateMock.getInstance().on('change', (args) => {
+										if (device3.id === args.deviceId) {
+											device3ChangeEventSpy(args);
+										}
+									});
+									await fakeDevice.getState(
+										device3,
+										device3.uuid,
+										stateVersion,
+									);
+									await waitFor({
+										checkFn: () => device3ChangeEventSpy.called,
+									});
+									device3ChangeEventSpy.resetHistory();
+									await expectDeviceHeartbeat(device3.id, {
+										db: DeviceOnlineStates.Online,
+										cache: DeviceOnlineStates.Online,
+									});
+
+									await setTimeout(heartbeatTimeoutChangeInterval);
+									await waitFor({
+										checkFn: () => device3ChangeEventSpy.called,
+									});
+									device3ChangeEventSpy.resetHistory();
+									await expectDeviceHeartbeat(device3.id, {
+										db: DeviceOnlineStates.Timeout,
+										cache: DeviceOnlineStates.Timeout,
+									});
+
+									// emulate the write cache having expired while the redis message queue hasn't been processed
+									// eg b/c the API/DB was down.
+									await redis.del(getHeartbeatWriteCacheKey(device3.id));
+								});
+
+								beforeEach(function () {
+									device3ChangeEventSpy.resetHistory();
+									delete tracker.states[device3.id];
+								});
+
+								itExpectsError(
+									'should turn Online and not consume the stale heartbeat redis queue message when a newer state GET has arrived',
+									async function () {
+										await fakeDevice.getState(
+											device3,
+											device3.uuid,
+											stateVersion,
+										);
+										await setTimeout(100);
+										// The write cache has expired so the DB's heartbeat gets updated
+										expect(device3ChangeEventSpy.called).to.be.true;
```

Comment:
```suggestion
										// The write cache has expired so the DB's heartbeat gets updated
										await waitFor({
											checkFn: () => device3ChangeEventSpy.called,
										});
```

Reply:
What the `await setTimeout(100);`s all over the test cases try to do, is to wait just enough time for the `'get-state'` event to be fired and processed by the API. The default max wait of `waitFor` is bigger than the mock device poll interval * jitter. As a result while in most test cases we want to wait for `it should shortly switch from timeout to online after a state get`, going for a plain waitFor would make it incorrectly accept a switch from timeout to offline as well.

I think I need to make a helper function w/ that setTimeout explains better why we need to do this...